sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py

```diff
@@ -0,0 +1,349 @@
+Running this example requires NumPy and SciPy
+"""
+import argparse
+import logging
+import string
+import uuid
+from collections import defaultdict
+
+import numpy as np
+from scipy.optimize import minimize
+
+import apache_beam as beam
+from apache_beam import pvalue
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+
+
+class Simulator(object):
+  """
+  Greenhouse simulation
+
+  Disclaimer: this code is an example and does not correspond to any real greenhouse simulation.
+  """
+
+  def __init__(self, quantities):
+    super(Simulator, self).__init__()
+    self.quantities = np.atleast_1d(quantities)
+
+    self.A = np.array([[3.0, 10, 30],
+                       [0.1, 10, 35],
+                       [3.0, 10, 30],
+                       [0.1, 10, 35]])
+
+    self.P = 1e-4 * np.array([[3689, 1170, 2673],
+                              [4699, 4387, 7470],
+                              [1091, 8732, 5547],
+                              [381, 5743, 8828]])
+
+    a0 = np.array([1.0, 1.2, 3.0, 3.2])
+    coeff = np.sum(np.cos(np.dot(np.atleast_2d(a0).T, self.quantities[None, :])), axis=1)
+    self.alpha = coeff / np.sum(coeff)
+
+  def simulate(self, xc):
+    # Map the input parameter to a cost for each crop.
+    f = -np.sum(self.alpha * np.exp(-np.sum(self.A * np.square(xc - self.P), axis=1)))
+    return np.square(f) * np.log(self.quantities)
+
+
+class CreateGrid(beam.PTransform):
+  """
+  A transform for generating the mapping grid.
+  """
+
+  class PreGenerateMappings(beam.DoFn):
+    """
+    ParDo implementation which splits of 2 records and generated a sub grid.
+
+    This facilitates parallellization of the grid generation.
+    Emits both the PCollection representing the subgrid, as well as the list
+    of remaining records. Both serve as an input to GenerateMappings
+    """
+
+    def process(self, element):
+      records = list(element[1])
+      # Split of 2 crops and pre-generate all combinations to facilitate parallellism
+      # No point splitting of a crop which can only be created in 1 greenhouse,
+      # split of crops with highest number of options.
+      best_split = np.argsort([-len(rec['transport_costs']) for rec in records])[:2]
+      rec1 = records[best_split[0]]
+      rec2 = records[best_split[1]]
+
+      # Generate & emit all combinations
+      for a in rec1['transport_costs']:
+        if a[1]:
+          for b in rec2['transport_costs']:
+            if b[1]:
+              combination = [(rec1['crop'], a[0]), (rec2['crop'], b[0])]
+              yield pvalue.TaggedOutput('splitted', combination)
+
+      # Pass on remaining records
+      remaining = [rec for i, rec in enumerate(records) if i not in best_split]
+      yield pvalue.TaggedOutput('combine', remaining)
+
+  class GenerateMappings(beam.DoFn):
+    """
+    ParDo implementation to generate all possible assignments of crops to greenhouses.
```

Comment:
You can start pydoc comments in the first line with """. (For examples see: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/complete/tfidf.py#L52)